### ISSUE TITLE
Executor.map should yield results not futures

### ISSUE THREAD

**Issue by minrk** [blaze/distributed#91]
Executor.map returns a list of Futures, but the standard Executor API suggests that it should return an iterable (e.g. generator) of actual results.

I came upon this when trying to demonstrate that Executors allow swapping of IPython/concurrent.futures and distributed Executors, only to discover that distributed's Executor isn't compatible with the rest.

**Comment by mrocklin** [comment]
Yes, this is one of our breaks from the concurrent.futures API, see  http://distributed.readthedocs.org/en/latest/related-work.html#id3

I justify this break within the context of distributed computing because moving data is expensive and because distributed supports futures within `submit/map` so `e.map(func, e.map(func, seq))` work the same in both cases.

Thoughts?  Justifiable or no?  Any thoughts on an appropriate compromise?

**Comment by minrk** [comment]
I'm honestly not sure. In IPython parallel, I have a similar issue, which I address by `View.map` returning a single AsyncMapResult, which is a *single* Future, iterating through which yields actual result.

But when I added the Executor layer, I made it a regular iterator yielding results.

I think it's fine to *have* a map that deviates from concurrent.futures, but to me, part of the point of Executor is the compatible API. In IPython parallel, the result is two `map`s: View.map that behaves how I like, and Executor.map, which is a thin shim around View.map that provides the common Executor API.

Since the difference is tiny:

```python
class CompatibleExecutor(distributed.Executor):
    """concurrent.futures-compatible Executor"""
    def map(self, *args, **kwargs):
        list_of_futures = super().map(*args, **kwargs)
        for f in list_of_futures:
            yield f.result() # yield results instead of Futures
```

I would either:

1. make this an option on Executor (e.g. `executor.compatibility_mode = True`) (I have no idea what the name should be)
2. include a small compatibility subclass whose goal is to be API-compatible with others, as in the snippet above
